### DISCUSSION TITLE
Float phase really expendable with LiFePO4? And: dbus-serialbattery never goes into float

### DISCUSSION THREAD

**Discussion by rperpe** [Louisvdw/dbus-serialbattery#632] [Q&A] [answered]
Hello

[Here](https://github.com/Louisvdw/dbus-serialbattery/discussions/170#discussioncomment-4213959) @Louisvdw stated LiFePO4 batteries don't really use float/absorption.

**Question 1: Is reducing of the charging voltage really expendable?** I'm in doubt since e.g. from [here](https://nordkyndesign.com/charging-marine-lithium-battery-banks/) I understand that bothering cells at high voltage with small currents can overcharge them.

This is what happens with my battery currently:
![image](https://github.com/Louisvdw/dbus-serialbattery/assets/52133854/e5651dd7-ead9-401e-bb12-91b1db8b7694)

I tried to activate dbus-serialbattery's [Float voltage emulation](https://louisvdw.github.io/dbus-serialbattery/general/features#float-voltage-emulation) which doesn't work for me. It looks like [this condition](https://github.com/Louisvdw/dbus-serialbattery/blob/v0.14.3/etc/dbus-serialbattery/battery.py#L141) is never evaluated to true. Root cause seems to be that my MultiPlus(es) measure voltage some ~0.1V to high (compared to the BMS) and therefore the battery pack never reaches `MAX_CELL_VOLTAGE * cell_count`.

**Question 2: Do you see any solution to this voltage measurement issue?** Would it perhaps make sense to make the current `MAX_CELL_VOLTAGE * cell_count` (float trigger voltage) configurable?

Many thanks

**Comment by mr-manuel** [comment]
Are you using version `v0.14.3`? Please try the latest beta version. This mechanism in `v0.14.3` has some errors. I was also improved in the new version.

I have to update the docs. See https://github.com/Louisvdw/dbus-serialbattery/blob/88e5eea93331ae99dc28fcf8b5a40cb866e432e2/etc/dbus-serialbattery/config.default.ini#L29-L63

This also could interrest you https://github.com/Louisvdw/dbus-serialbattery/issues/458#issuecomment-1537219671

**Reply by rperpe** [reply]
Hello @mr-manuel, many thanks for your reply.

Yes I'm on latest, v0.14.3.

I'll give the beta a try this weekend but I'm not very confident that it will change regarding my issue, since [this condition's](https://github.com/Louisvdw/dbus-serialbattery/blob/v1.0.20230508beta/etc/dbus-serialbattery/battery.py#L312) problematic part seems to remain the same:

`utils.MAX_CELL_VOLTAGE * self.cell_count <= voltageSum`

Since my cells (with MAX_VOLTAGE = 3.5V and 17 cells) only go until ~3.49V each, `3.5 * 17 <= 3.49 * 17` becomes never true.

Am I wrong?

**Reply by mr-manuel** [reply]
1. What have you selected as battery monitor? The multiplus should not set the voltage but the driver so this should not occur.
2. Which Venus OS are you using?
3. How many batteries do you have attached?
4. Have you enabled DVCC?
5. What does the MultiPlus display for a voltage?
6. What does the SerialBattery driver display for a voltage?

I would recommend you to try the linear mode with the beta version.

https://github.com/Louisvdw/dbus-serialbattery/blob/88e5eea93331ae99dc28fcf8b5a40cb866e432e2/etc/dbus-serialbattery/config.default.ini#L35

**Reply by rperpe** [reply]
I'll give the beta and the linear mode a try.

Regarding your questions:
1. Battery monitor is Victron SmartShunt since it is most accurate. To me it looks like the driver correctly sets the MultiPlus' voltage to 17*3.5V=59.5V, at least this is the voltage which is visible at end of charging in MultiPlus ... -> DC Voltage and SerialBattery(Jkbms) -> Parameters. Venus OS' graphical overview shows "ext. control" while charging.
BMS' and SmartShunt's voltage measurement remain ~0.1V below (and therefore I believe the driver never goes into float).
2. Venus OS 3.00~37 and 3.00~39 since today.
3. Currently only one battery pack with 17 cells.
4. Yes, DVCC is enabled (all other DVCC options below are disabled), Controlling BMS in DVCC is "SerialBattery(Jkbms)".
5. MultiPlus shows 59.5V when fully charged, actual voltage remains slightly below.
6. ~59.35V

**Reply by mr-manuel** [reply, marked as answer]
Ok, now I understood.

The `dbus-serialbattery` sets correctly the CVL voltage, but since you are using the SmartShunt as battery monitor, which displays `+0.15V`, it stops charging before. You will need an option to add a voltage difference, e.g. if you have a voltage drop on the cable.

Could you measure the voltage with a multimeter on the SmartShunt's side and on the battery side?

I added `VOLTAGE_DROP` in the `config.default.ini` so you can add it to your `config.ini` as you need.

Install a specific version and use this release: https://github.com/mr-manuel/venus-os_dbus-serialbattery/releases/tag/v1.0.20230509beta

**Reply by rperpe** [reply]
I successfully installed the `v1.0.20230509beta`, it really contains many great improvements. Many thanks @mr-manuel for this release and the phenomenally fast solution to this issue!

I'm not 100% sure if the problem exists only cause of existence of the SmartShunt in my installation. I'll monitor in the next days, do some measurements and tests without the shunt and - if there are findings - will open a PR to adapt the docs in config.ini.default.

**Reply by mr-manuel** [reply]
The problem exists because you are measuring the voltage in two different places. Since due to the cable there is a small voltage drop. If you select the BMS for DVCC and battery monitor the voltages matches, so no problem. Else your battery monitor should decide the voltage to charge with. Then also the problem wouldn't exist.

**Reply by rperpe** [reply]
I see, usually the BMS gets set as a battery monitor, not the MultiPlus'es integrated battery monitor.

✔️ lowering CVL to (cell count * float voltage) seems to have worked today 👍👍

**Comment by rperpe** [comment]
Is it OK to ask two more questions regarding the [v1.0.20230509beta release](https://github.com/mr-manuel/venus-os_dbus-serialbattery/releases/tag/v1.0.20230509beta) from @mr-manuel in this answered discussion?

Ahead: VOLTAGE_DROP feature seems to work fine, would be nice if this gets merged into upstream 👍.

Q1: In config.ini I configured `MAX_VOLTAGE_TIME_SEC = 1800`, shouldn't `MAX_VOLTAGE` then be held for at least half an hour? The driver lowers to `FLOAT_CELL_VOLTAGE` much earlier, on this day after ~13min:
![image](https://github.com/Louisvdw/dbus-serialbattery/assets/52133854/be18c562-b072-48ad-906e-75807044e604)

Q2: When should CVL be reset from `FLOAT_CELL_VOLTAGE` back to `MAX_VOLTAGE`? Since yesterday this didn't happen (SoC was around ~40% this morning) and therefore I restarted the driver today (had to leave and couldn't monitor what will happen as soon as `FLOAT_CELL_VOLTAGE` is hit).

I tried to track with the Python code, but unfortunately I don't understand enough of it. Sent you a coffee via PayPal.

**Reply by mr-manuel** [reply]
It will get merged before next release.

Q1 & Q2: https://github.com/mr-manuel/venus-os_dbus-serialbattery/blob/72a40175f00b9213a022918677345d41eb69e2b5/etc/dbus-serialbattery/config.default.ini#L44-L80

Q2: It should. What do you see under the parameters page? Please post a screenshot, the next time the voltage drops again. I can't test it since I have no good weather since a week.

Thank you!

**Reply by rperpe** [reply]
Ah, I see (didn't realize that behaviour isn't the same for linear and step mode).

Weather is also questionable here 😄. However it was enough to reach 100% SoC. dbus-serialbattery charged to max. voltage (green):

![image](https://github.com/Louisvdw/dbus-serialbattery/assets/52133854/52c69afa-837d-40dc-9ecc-b2d59ad03ce0)

Then it went up&down (all correct, due to clouds and usage) and CVL was lowered (yellow).

Now SoC (all three available SoCs) is below 90%...
![image](https://github.com/Louisvdw/dbus-serialbattery/assets/52133854/3429d416-3012-489d-9328-9aa77b1684cc)
![image](https://github.com/Louisvdw/dbus-serialbattery/assets/52133854/040e27e2-7ed5-4cf6-bce9-e1244fa36745)
![image](https://github.com/Louisvdw/dbus-serialbattery/assets/52133854/84078293-33ab-4086-9e22-ec6e577feae4)

...but still CVL is on float value:
![image](https://github.com/Louisvdw/dbus-serialbattery/assets/52133854/148280c0-a3e6-442e-9e6f-88f75775b7bb)

Reset config is untouched:
![image](https://github.com/Louisvdw/dbus-serialbattery/assets/52133854/43a8d6bc-575d-40c0-959d-96bc73cb0519)

Do you have an idea what's going wrong so that it doesn't reset CVL to max. voltage?

**Reply by mr-manuel** [reply]
Oh, it's not included in `v1.0.20230509beta`. Please install [`v1.0.20230515beta`](https://github.com/mr-manuel/venus-os_dbus-serialbattery/releases/tag/v1.0.20230515beta).

**Reply by rperpe** [reply]
Will test this, many thanks!

**Reply by rperpe** [reply]
[v1.0.20230515beta](https://github.com/mr-manuel/venus-os_dbus-serialbattery/releases/tag/v1.0.20230515beta) works fine 👍
